Fix  invariant error in fee-sized `VaultWithdraw`

Bronek opened the pull request:
<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

This fixes an invariant error where the amount withdrawn is equal to the transaction fee.

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

The invariants for `VaultWithdraw` enforce that the account receiving the funds must have their balance changed (increased). This won't happen when the amount withdrawn is exactly same as fee. The submitter account balance won't change, because the deduced transaction fee is exactly refilled with the withdrawn amount.

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [x] Bug fix (non-breaking change which fixes an issue)

gregtatcam commented on src/xrpld/app/tx/detail/InvariantCheck.cpp:
```diff
@@ -2297,7 +2297,7 @@ ValidVault::visitEntry(
 
     uint256 const key = (before ? before->key() : after->key());
-    if (sign && balance != zero)
+    if (sign != 0)
```
Why is the balance check removed?

Bronek replied:
Because zero balance change, as long as it came from an object actually being updated, contains valid and useful information which the Vault invariants rely on.

The bug here is that the withdrawal amount is exactly equal to the fee, so the transaction does not change its TX submitter XRP balance. If we keep `balance != zero` condition, we lose the information that the balance was updated but the new XRP amount happened to be the same as the old amount. That information is in turn required by the invariant, and if not provided we have invariant violation `withdrawal must change one destination balance`

ximinez replied:
> The bug here is that the withdrawal amount is exactly equal to the fee, so the transaction does not change its TX submitter XRP balance.

What happens if the fee is larger than the withdraw amount? The tx submitter's balance would decrease. That's still valid.

ximinez replied:
I was looking again at `accountDeltaAssets` (in a couple of places), and noticed that the lambda adds in the amount of the `fee` paid for `tx[sfAccount]`. That is going to give the wrong result for a delegated transaction, since the delegate pays the fee.

Fortunately, you only need to check for the presence of `sfDelegate` to determine who paid the fee. You may also want to factor those two lambdas into one that can be called from the two locations so you don't have to worry about drift.

Bronek replied:
Moved to single location in f36dd22cae1fdc42f0e5cfd8638a634e6855c425 however I am not sure whether to keep this check

```c++
        // Delegated transaction; no need to compensate for fees
        if (auto const delegate = tx[~sfDelegate];
            delegate.has_value() && *delegate != tx[sfAccount])
            return ret;
```

... since this is currently dead code, with no way to test.

Bronek replied:
... actually I was wrong, this can be tested - see c3b0562c7b . Although I am not sure how faithfully such a test represents an actual delegated transaction.

gregtatcam commented on src/test/app/Vault_test.cpp:
```diff
@@ -304,6 +304,41 @@ class Vault_test : public beast::unit_test::suite
+                testcase(prefix + " deposit/withdrawal same as fee");
+                auto const amount = env.current()->fees().base;
+
+                auto tx = vault.deposit(
```
Shouldn't per PR description all tx fail if the tx amount is the same as the fee?

Bronek replied:
no, it should not - we want this transaction to succeed because there's nothing inherently invalid about withdrawing an amount which matches exactly the fee.

gregtatcam replied:
Got it! I misread the description.

gregtatcam reviewed (approved):
LGTM 👍

ximinez commented on src/test/app/Invariants_test.cpp:
Spelling: `Adjustments`

Bronek replied:
Thanks, fixed in 63014310f8

ximinez commented on src/test/app/Vault_test.cpp:
```diff
@@ -4795,6 +4831,136 @@ class Vault_test : public beast::unit_test::suite
+                                   Account const& account) -> long {
+            auto sle = env.le(keylet::account(account.id()));
+            assert(sle != nullptr);
+            return sle->getFieldAmount(sfBalance).xrp().drops();
```
This `assert` could be a `BEAST_EXPECT`.
```
if(!BEAST_EXPECT(sle != nullptr))
    return 0; // or some weird value like -1
```
Don't bother checking the result of `xrpBalance`, because the test has failed, so it'll cascade.

Bronek replied:
Thanks, done in 328369b57b155039b3da96842ee9c6203ca1959d

ximinez commented on src/xrpld/app/tx/detail/InvariantCheck.cpp:
```diff
@@ -2297,7 +2297,7 @@ ValidVault::visitEntry(
 
     uint256 const key = (before ? before->key() : after->key());
-    if (sign && balance != zero)
+    if (sign != 0)
```
The description of `balance` makes clear that it's a delta. But looking at a line like this in isolation, it's easy to forget. I think it would make sense to rename it to indicate more accurately what it is. Something like `delta`, `balanceDelta`, `balanceChange`, `balanceDifference`....

Bronek replied:
Thanks, done in 328369b57b155039b3da96842ee9c6203ca1959d

ximinez reviewed (changes requested):
Just a couple of small suggestions

ximinez commented on src/test/app/Vault_test.cpp:
```diff
@@ -4846,11 +4846,15 @@ class Vault_test : public beast::unit_test::suite
-                                   Account const& account) -> long {
+        // Special "error" value to recognize in tests if things fail.
+        // Accidentally it is also 2^31 - 1 and a Mersenne prime.
+        static constexpr long fail = 2'147'483'647;
```
I'm not usually a fan of "magic numbers", but since this is test code, I'm ok with it. If you want to rewrite the lambda to return a `std::optional<long>`, I wouldn't complain about it, but I know you're eager to get this merged.

ximinez commented on src/xrpld/app/tx/detail/InvariantCheck.cpp:
```diff
@@ -2230,13 +2230,15 @@ ValidVault::visitEntry(
     // invariants can validate that the change in account balances matches the
     // change in vault balances, stored to deltas_ at the end of this function.
-    Number balance{};
+    // This is done even if balanceDelta is zero, but an object was updated.
+    Number balanceDelta{};
 
-    // By default do not add anything to deltas
+    // Append to deltas if sign is non-zero, i.e. an object of an interesting
+    // type has been updated - even if the update did not change the balance.
     std::int8_t sign = 0;
```
These comments are contradictory, and the code behavior is to only append if the sign is not zero, so the first one is wrong. If the update did not change the balance, the sign will be zero, and this the value will not be appended.

Bronek replied:
> If the update did not change the balance, the sign will be zero

I think that's incorrect. SAV transactors (as all others) will call `view().update(sle)` unconditionally, whether or not the balance has changed, and we actually rely on this behaviour. The specifics in this particular PR: withdraw increases the submitter balance by the exact same amount of XRP as the transaction fee, hence submitters `sfBalance` will not change but the account root object will be still updated. We also handle even more extreme case, when the submitter withdraws less XRP than the transaction fee, hence their balance will decrease (I guess I need to add test for this one).

In the context of this block of code, `sign` will be set to -1 in `case ltACCOUNT_ROOT:` but since both `before->getFieldAmount(sfBalance)` and `after->getFieldAmount(sfBalance)` yield the same number, `balanceDelta` will be zero.

Bronek replied:
I guess I need to change the comment to make it clear that updating an object happens whether or not its balance has changed, and that includes account root objects.

Bronek replied:
hope the comment in 26b57ef392 makes this clear.

ximinez replied:
> In the context of this block of code, `sign` will be set to -1 in `case ltACCOUNT_ROOT:` but since both `before->getFieldAmount(sfBalance)` and `after->getFieldAmount(sfBalance)` yield the same number, `balanceDelta` will be zero.

Oh, I'm foolish. I was thinking of the `sign()` of `balanceDelta`, which _is_ zero in this case, but `sign` is set manually and independently of `balanceDelta`.

ximinez reviewed (approved):
I wouldn't _complain_ if you renamed the previously discussed `sign` variable, but this is good to go in it's current state.

(`deltaModifier`?)

bthomee committed:
fix: invariant error in fee-sized `VaultWithdraw`

This changes fixes an invariant error where the amount withdrawn is equal to the transaction fee.
